File: scripts/check-workflow-pins.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "find"
require "psych"

def mapping_pairs(mapping)
  mapping.children.each_slice(2)
end

def scalar_key?(node, name)
  node.is_a?(Psych::Nodes::Scalar) && node.value == name
end

def add_uses_entry(key, value, path, entries)
  entries << [path, key.start_line + 1, value.is_a?(Psych::Nodes::Scalar) ? value.value : nil]
end

def scan_step(step, path, entries)
  return unless step.is_a?(Psych::Nodes::Mapping)

  mapping_pairs(step).each do |key, value|
    add_uses_entry(key, value, path, entries) if scalar_key?(key, "uses")
  end
end

def scan_job(job, path, entries)
  return unless job.is_a?(Psych::Nodes::Mapping)

  mapping_pairs(job).each do |key, value|
    if scalar_key?(key, "uses")
      add_uses_entry(key, value, path, entries)
    elsif scalar_key?(key, "steps") && value.is_a?(Psych::Nodes::Sequence)
      value.children.each { |step| scan_step(step, path, entries) }
    end
  end
end

def scan_document(document, path, entries)
  root = document.children.first
  return unless root.is_a?(Psych::Nodes::Mapping)

  mapping_pairs(root).each do |key, value|
    next unless scalar_key?(key, "jobs") && value.is_a?(Psych::Nodes::Mapping)

    mapping_pairs(value).each { |_job_name, job| scan_job(job, path, entries) }
  end
end

def scan_file(path, entries)
  stream = Psych.parse_stream(File.read(path), filename: path)
  stream.children.each { |document| scan_document(document, path, entries) }
rescue Psych::Exception => error
  warn "Failed to parse workflow file #{path}: #{error.message}"
  exit 2
rescue SystemCallError => error
  warn "Failed to read workflow file #{path}: #{error.message}"
  exit 2
end

workflow_directory = ARGV.fetch(0)
entries = []

begin
  Find.find(workflow_directory) do |path|
    next unless File.file?(path) && path.match?(/\.ya?ml\z/)

    scan_file(path, entries)
  end
rescue SystemCallError => error
  warn "Failed to scan workflow files: #{error.message}"
  exit 2
end

invalid_entries = entries.reject do |_path, _line, value|
  value&.start_with?("./", "docker://") || value&.match?(/\A[^[:space:]@]+@[0-9a-fA-F]{40}\z/)
end

unless invalid_entries.empty?
  warn "Workflow actions must use full 40-character commit SHAs:"
  invalid_entries.each do |path, line, value|
    warn "#{path}:#{line}: uses: #{value}"
  end
  exit 1
end

File: scripts/check-workflow-pins.mjs
#!/usr/bin/env node
import { readdirSync, readFileSync, statSync } from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { parseAllDocuments } from 'yaml'

class WorkflowScanError extends Error {}

const PINNED = /^[^\s@]+@[0-9a-fA-F]{40}$/
const SELF_REPOSITORY = /^\$\/[^\s@]+$/

function ensureAcyclic(node, ancestors = new Set()) {
  if (node === null || typeof node !== 'object') {
    return
  }
  if (ancestors.has(node)) {
    throw new WorkflowScanError('cyclic YAML aliases are not supported')
  }
  ancestors.add(node)
  if (Array.isArray(node)) {
    for (const value of node) {
      ensureAcyclic(value, ancestors)
    }
  } else {
    for (const value of Object.values(node)) {
      ensureAcyclic(value, ancestors)
    }
  }
  ancestors.delete(node)
}

function isMapping(value) {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function addUsesEntry(value, file, entries) {
  entries.push([file, typeof value === 'string' ? value : null])
}

function scanStep(step, file, entries) {
  if (isMapping(step) && 'uses' in step) {
    addUsesEntry(step.uses, file, entries)
  }
}

function scanJob(job, file, entries) {
  if (!isMapping(job)) {
    return
  }
  if ('uses' in job) {
    addUsesEntry(job.uses, file, entries)
  }
  if (Array.isArray(job.steps)) {
    for (const step of job.steps) {
      scanStep(step, file, entries)
    }
  }
}

function scanDocument(document, file, entries) {
  if (!isMapping(document) || !isMapping(document.jobs)) {
    return
  }
  for (const job of Object.values(document.jobs)) {
    scanJob(job, file, entries)
  }
}

function fail(stage, file, error) {
  process.stderr.write(`Failed to ${stage} workflow file ${file}: ${error.message}\n`)
  process.exit(2)
}

function scanFile(file, entries) {
  let source
  try {
    source = readFileSync(file, 'utf8')
  } catch (error) {
    fail('read', file, error)
  }

  // Alias resolution belongs to parsing: an unresolved alias is a broken
  // document, while a cycle only shows up once the document is materialized.
  let documents
  let document
  try {
    documents = parseAllDocuments(source, { merge: true })
    for (const parsed of documents) {
      if (parsed.errors.length > 0) {
        throw parsed.errors[0]
      }
    }
    document = documents.length === 1 ? documents[0].toJS({ maxAliasCount: -1 }) : undefined
  } catch (error) {
    fail('parse', file, error)
  }

  try {
    if (documents.length !== 1) {
      throw new WorkflowScanError('multiple YAML documents are not supported')
    }
    ensureAcyclic(document)
    scanDocument(document, file, entries)
  } catch (error) {
    fail('scan', file, error)
  }
}

function workflowFiles(directory) {
  const files = []
  const pending = [directory]
  while (pending.length > 0) {
    const current = pending.pop()
    const info = statSync(current)
    if (info.isDirectory()) {
      for (const entry of readdirSync(current).sort()) {
        pending.push(path.join(current, entry))
      }
    } else if (info.isFile() && /\.ya?ml$/.test(current)) {
      files.push(current)
    }
  }
  return files.sort()
}

const directory = process.argv[2]
if (directory === undefined) {
  process.stderr.write('usage: check-workflow-pins.mjs <workflow-directory>\n')
  process.exit(2)
}

const entries = []
let files
try {
  files = workflowFiles(directory)
} catch (error) {
  process.stderr.write(`Failed to scan workflow files: ${error.message}\n`)
  process.exit(2)
}
for (const file of files) {
  scanFile(file, entries)
}

const invalid = entries.filter(([, value]) => {
  if (typeof value !== 'string') {
    return true
  }
  return !(
    value.startsWith('./') ||
    value.startsWith('docker://') ||
    SELF_REPOSITORY.test(value) ||
    PINNED.test(value)
  )
})

if (invalid.length > 0) {
  process.stderr.write('Workflow actions must use full 40-character commit SHAs:\n')
  for (const [file, value] of invalid) {
    process.stderr.write(`${file}: uses: ${value ?? 'uses value is not a scalar'}\n`)
  }
  process.exit(1)
}
